### PR TITLE
Xy calibration goodies2

### PR DESCRIPTION
**Pull request by vedina**
- test for laser zeroing verifies if the Si peaks is at the correct position
- additional method for peak matching using scipy linear_assignment
- CalibrationModel and XCalibrationComponent options to select the match_method 
- demo notebook

### CODE REVIEW

**Comment by vedina** [comment]
@georgievgeorgi  @nicocopez if you could go through the demo notebook and tell 
- what is not clear
- what you want to change and wondering how , etc 
then I will add explanations (or additional code)

File: src/ramanchada2/io/HSDS.py [hunk 1]
```diff
@@ -67,18 +67,22 @@ def write_nexus(filename: str,
 class DatasetExistsError(Exception):
     pass
 
+def sanitize_key(key: str) -> str:
+    return ''.join(c if ord(c) < 128 else '_' for c in key)
```
Comment (kerberizer) [inline src/ramanchada2/io/HSDS.py:71]: Instead of `ord(c) < 128`, maybe use [`c.isascii()`](https://docs.python.org/3/library/stdtypes.html#str.isascii) or even `c.isalnum()` if the key will be only alphanumeric?

File: src/ramanchada2/protocols/calibration.py [hunk 2]
```diff
@@ -582,20 +614,22 @@ def __init__(self, laser_wl: int):
         Example:
         ```python
         # Create an instance of CalibrationModel
-        calmodel = CalibrationModel(laser_wl=785)
-        calmodel.derive_model_x(
+        import ramanchada2 as rc2
+        import ramanchada2.misc.constants as rc2const
+        from ramanchada2.protocols.calibration import CalibrationModel
```
Comment (kerberizer) [inline src/ramanchada2/protocols/calibration.py:619]: Why do we put these imports here, and why do we import the whole rc2?

File: src/ramanchada2/misc/utils/argmin2d.py [hunk 3]
```diff
@@ -145,3 +146,162 @@ def match_peaks(spe_pos_dict, ref):
...
+    order_penalty = order_weight * abs(p1[0] - p2[0])
+    return order_penalty
+
+
+def cost_function(p1, p2, order_weight=1.0, priority_weight=.1):
+    """
+    Modified cost function with an order preservation penalty and priority weighting.
+    - `order_weight` increases penalty for large differences in the x-axis values.
+    - `priority_weight` decreases the cost for higher values in the y-axis for set_b points.
+    """
```
Comment (kerberizer) [inline src/ramanchada2/misc/utils/argmin2d.py:161]: It might be nice to follow the agreed upon docstring format: https://docs.python.org/3/library/stdtypes.html#str.isascii

File: src/ramanchada2/misc/utils/argmin2d.py [hunk 4]
```diff
@@ -145,3 +146,162 @@ def match_peaks(spe_pos_dict, ref):
...
+    -------
+    matched_peaks : (matched_peaks_a,matched_peaks_b,matched_distances, df)
+
+    Examples:
+    ---------
+    >>> spectrum_a = {100: 10, 105: 20, 110: 15}
+    >>> spectrum_b = {102: 12, 106: 22, 111: 16}
+    >>> match_peaks(spectrum_a, spectrum_b)
+
+    """
```
Comment (kerberizer) [inline src/ramanchada2/misc/utils/argmin2d.py:239]: This seems more like numpy docstring syntax.

**Review by kerberizer — COMMENTED** [review]
Have not looked into the domain specific code for obvious reasons.

File: src/ramanchada2/io/HSDS.py (hunk 1 above)
Comment (vedina) [reply src/ramanchada2/io/HSDS.py:71]: isascii should be fine. 
there are some .spc with metadata keys with specific unicode that are breaking hdf5 writing

File: src/ramanchada2/protocols/calibration.py (hunk 2 above)
Comment (vedina) [reply src/ramanchada2/protocols/calibration.py:619]: this is example in pydoc

File: src/ramanchada2/io/HSDS.py (hunk 1 above)
Comment (vedina) [reply src/ramanchada2/io/HSDS.py:71]: we don't need alphanum only, keys can contain brackets, dots, etc

File: src/ramanchada2/protocols/calibration.py (hunk 2 above)
Comment (kerberizer) [reply src/ramanchada2/protocols/calibration.py:619]: Ugh, my bad, thanks!

File: src/ramanchada2/io/HSDS.py (hunk 1 above)
Comment (kerberizer) [reply src/ramanchada2/io/HSDS.py:71]: If we know what the valid characters are, maybe test for them? Because currently we would also allow some non-printable characters, e.g. ^G, BELL.

File: src/ramanchada2/misc/utils/argmin2d.py
Comment (georgievgeorgi) [inline src/ramanchada2/misc/utils/argmin2d.py:1]: Would be better to have implementations not related to argmin2d in a separate file.

Comment (vedina) [reply src/ramanchada2/misc/utils/argmin2d.py:1]: Agree !

File: src/ramanchada2/protocols/calibration.py
Comment (georgievgeorgi) [inline src/ramanchada2/protocols/calibration.py:1]: the one-class-per-file rule will increase readability

Comment (vedina) [reply src/ramanchada2/protocols/calibration.py:1]: I thought Python is against that rule

File: tests/protocols/test_calibrationmodel.py [hunk 5]
```diff
@@ -86,22 +101,48 @@ def setup_module():
...
         _units = rf"$\mathrm{{[cm^{-1}]}}$"
     assert _units == "nm"
     ax.set_xlabel(_units)
-    # print(setup_module.calmodel.components[1])
+    spe_test = spe_sil_calib  # .trim_axes(method='x-axis',boundaries=(si_peak-50,si_peak+50))
+    find_kw = {"wlen": 200, "width": 1, "sharpening": None}
+    find_kw["prominence"] = spe_test.y_noise_MAD() * setup_module.calmodel.prominence_coeff
+    cand = spe_test.find_peak_multipeak(**find_kw)
+    fit_kw = {}
+    fitres = spe_test.fit_peak_multimodel(profile=setup_module.si_profile, candidates=cand, **fit_kw, no_fit=False)
```
Comment (georgievgeorgi) [inline tests/protocols/test_calibrationmodel.py:123]: `**fit_kw` here makes no sense after `fit_kw = {}`

File: src/ramanchada2/misc/utils/argmin2d.py
Comment (vedina) [reply src/ramanchada2/misc/utils/argmin2d.py:1]: Done

File: tests/protocols/test_calibrationmodel.py [hunk 6]
```diff
@@ -86,22 +101,48 @@ def setup_module():
...
+    spe_test_necalibrated_only.plot(ax=axsifit, label="Si (Ne calibrated only)", fmt=":")
+
+    df = fitres.to_dataframe_peaks().sort_values(by="height", ascending=False)
+    print(df.iloc[0]["center"])
+
+    print(df.sort_values(by="amplitude", ascending=False).head())
+
+    plt.grid()
     plt.tight_layout()
     plt.savefig("test_calmodel_{}.png".format("laser_zeroing"))
```
Comment (georgievgeorgi) [inline tests/protocols/test_calibrationmodel.py:143]: Do you really need to generate plots in the tests? They are slow and do not believe that they are going to be used somehow as part of the tests.

File: src/ramanchada2/misc/utils/matchsets.py [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+    sort_indices = np.argsort(x_spe)
+    return (x_spe[sort_indices], x_reference[sort_indices], x_distance[sort_indices], df)
+
+
+def cost_function_position(p1, p2, order_weight=1.0, priority_weight=1.0):
+    order_penalty = order_weight * abs(p1[0] - p2[0])
+    return order_penalty
+
+
+def cost_function(p1, p2, order_weight=1.0, priority_weight=.1):
```
Comment (georgievgeorgi) [inline src/ramanchada2/misc/utils/matchsets.py:76]: It is not clear what `p1` and `p2` are. One can search in the source and see that these are supposed to be two element lists having positions and intensities. Having the code written this way is difficult to follow and makes it more difficult to focus on the algorithms.

File: src/ramanchada2/misc/utils/matchsets.py [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+    """
+    Modified cost function with an order preservation penalty and priority weighting.
+    - `order_weight` increases penalty for large differences in the x-axis values.
+    - `priority_weight` decreases the cost for higher values in the y-axis for set_b points.
+    """
+    order_penalty = order_weight * abs(p1[0] - p2[0])
+    priority_bonus = priority_weight * p2[1]  # Rewards points in set_b with higher second dimension values
```
Comment (georgievgeorgi) [inline src/ramanchada2/misc/utils/matchsets.py:83]: This is not always desirable. We may have a fake high-amplitude reference/peak

File: src/ramanchada2/misc/utils/matchsets.py [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+    >>> spectrum_a = {100: 10, 105: 20, 110: 15}
+    >>> spectrum_b = {102: 12, 106: 22, 111: 16}
+    >>> match_peaks(spectrum_a, spectrum_b)
+
+    """
+    cost_matrix = cost_matrix_peaks(spectrum_a_dict, spectrum_b_dict, threshold_max_distance=threshold_max_distance, cost_func=cost_function if cost_func is None else cost_func)
+
+    # Use the Hungarian algorithm to find the optimal assignment
+    try:
+        row_ind, col_ind = linear_sum_assignment(cost_matrix)
```
Comment (georgievgeorgi) [inline src/ramanchada2/misc/utils/matchsets.py:164]: If we ignore the amplitude values `linear_sum_assignment` provides worse result than the already implemented `argmin2d`.

Consider this example:
```python
from scipy.optimize import linear_sum_assignment

import ramanchada2 as rc2
import numpy as np
import matplotlib.pyplot as plt

x = np.array([1, 1.5, 2, 2.5, 3, 6, 7, 7.1, 9, 10, 11, 12])
y = np.array([1, 2, 3, 4, 5, 6, 7, 8.])

lsa_x_idx, lsa_y_idx = np.array(linear_sum_assignment(np.abs(np.subtract.outer(x, y))))
gg_x_idx, gg_y_idx = rc2.misc.utils.find_closest_pairs_idx(x, y)

line_kw = dict(alpha=.5)
for i in x:
    plt.axvline(i, **line_kw)
    
for i in y:
    plt.axhline(i, **line_kw)
    
plt.plot(x[lsa_x_idx], y[lsa_y_idx], 'x', markersize=10, mew=3, label='linear_sum_assignmet')
plt.plot(x[gg_x_idx], y[gg_y_idx], '+', markersize=10, mew=3, label='gg argmin2d')
plt.legend(loc='lower right')
plt.xlabel('locations 1')
plt.ylabel('locations 2')
```

File: src/ramanchada2/misc/utils/matchsets.py [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+from sklearn.metrics.pairwise import euclidean_distances
+from scipy.optimize import linear_sum_assignment
+import numpy as np
+
+def match_peaks_cluster(spe_pos_dict, ref):
+    # Min-Max normalize the reference values
+    min_value = min(ref.values())
+    max_value = max(ref.values())
+    if len(ref.keys()) > 1:
+        normalized_ref = {key: (value - min_value) / (max_value - min_value) for key, value in ref.items()}
```
Comment (georgievgeorgi) [inline src/ramanchada2/misc/utils/matchsets.py:13]: should not subtract `min_value` as this way the smallest peak is getting lost. min/max makes more sense in a spectrum, not individual peaks

**Comment by vedina** [comment]
- Resolved the review comments
- refactoring of the protocols/calibration into (semi)separate classes
- optional plots in tests (disabled by default)
- fixed bug in match_peaks_cluster in interpretation of clustering results
- added options for match methods
- added options for interpolators - we now have RBFInterpolator, PchipInterpolator, CubicSpline . Of these, only PchipInterpolator guarantees monotonic calibration curve, and when interpolating.  But Ne peaks don't cover the entire range, so we need (some) extrapolation. Hence added extrapolation option.
- option what to do if the results of applying the calibration curve is nonmontonic - ignore, set nans, drop nans, raise error

@georgievgeorgi @kerberizer
